Add subscription info to invite emails & endpoint for sending "trial ending soon" emails

emma-sg opened the pull request:
Closes #2685 
Closes #2798

## Changes
- Adds subscription info (specifically days until trial ends) to invite emails
- Adds new endpoint for sending a "trial ending soon" email to all admins based on subscription id

## Testing

Tested in dev alongside https://github.com/webrecorder/cashew/pull/57! All seems to work.

Tested locally by doing the following:
1. Rebuild the backend and emails images with `scripts/build-backend.sh` and `scripts/build-emails.sh`
2. Set a value for `BTRIX_SUBS_APP_API_KEY` in the `btrix-subs-app-secret` secret by doing the following:
    1. Choose a secret and write it down — I just used `uuidgen` to generate one
    2. Base64-encode the secret, I used `cat '<my secret>' | base64`
    3. Run `kubectl edit secrets btrix-subs-app-secret` and paste your encoded secret in the `BTRIX_SUBS_APP_API_KEY` field, and save and exit
3. Redeploy your local Browsertrix with `helm upgrade --install -f ./chart/values.yaml -f ./chart/local.yaml btrix ./chart/ && kubectl wait --for=condition=ready pod --all --timeout=300s`
4. Manually modify the `subscription` field on the org document in MongoDB
    I used the following data in MongoDB Compass connected to my local Browsertrix instance: 
    ```json
    "subscription": {
      "subId": "stripe:sub_abc123",
      "planId": "starter",
      "status": "trialing",
      "futureCancelDate": {
        "$date": "2025-08-14T21:14:29.000Z"
      },
      "readOnlyOnCancel": false
    },
    ```
    Pick some date a bit into the future for the `futureCancelDate`.
5. Add an admin user or two to an org, and follow the sign-up process
    - **Verify that the "Manage Your Billing and Plan" section on the sign-up emails mentions when the trial ends**
6. Set up smtp4dev or something similar to receive emails sent from Browsertrix (see [Testing Email Sending from Browsertrix](https://docs.browsertrix.com/develop/emails/#testing-email-sending-from-browsertrix))
7. Run the following to get Browsertrix to send trial ending soon emails to all the admins of your org:
    ```sh
    curl -X 'POST' \
      'http://browsertrix-cloud-backend.default.svc.cluster.local:8000/api/subscriptions/trial-end-reminder' \
      -H 'Content-Type: application/json' \
      -d '{
      "subId": "stripe:sub_abc123",
      "behavior_on_trial_end": "cancel"
    }' -H 'Authorization: bearer <your secret from step 2.i>'
    ```
    Try with `behavior_on_trial_end` set to both `"cancel"` and `"continue"`.
    - **Verify that the dates & durations are accurate**
    - **Verify that the links to the org all work**
    - **Verify that all admins receive emails**

## Follow-ups

Once this is deployed, we should turn off trial end reminder emails from Stripe.

emma-sg committed:
add subscription info to invite emails

emma-sg committed:
add endpoint to send "trial will end" email

emma-sg committed:
format

emma-sg committed:
format & fix lint issues

emma-sg committed:
revert unnecessary change to email

emma-sg committed:
fix date calculation

emma-sg committed:
fail if not admins found for org

also change other lookup errors to 500s

emma-sg committed:
trim trailing slash from org urls

emma-sg committed:
use same day counting logic in invites & trial_ending_soon emails

emma-sg committed:
remove unused import

emma-sg committed:
allow `null` values for trial_end_date

emma-sg committed:
simplify email template dates with `z.coerce.date()`

tw4l commented on backend/btrixcloud/invites.py:
```diff
@@ -199,7 +212,7 @@ async def invite_user(
             oid=oid,
             created=dt_now(),
-            role=invite.role if hasattr(invite, "role") else None,
+            role=invite.role if hasattr(invite, "role") else UserRole.VIEWER,
```
Nice catch

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -1940,6 +1948,9 @@ class Subscription(BaseModel):
     futureCancelDate: Optional[datetime] = None
+    # pylint: disable=C0301
+    "When in a trial, future cancel date is the trial end date; when not in a trial, future cancel date is the date the subscription will be canceled, if set"
+
```
```suggestion
    # When in a trial, future cancel date is the trial end date.
    # When not in a trial, future cancel date is the date the subscription will
    # be cancelled, if set.
    futureCancelDate: Optional[datetime] = None

```

Think a regular comment here might be more appropriate. Unless there's something I'm missing?

emma-sg replied:
At least in VSCode and other editors that use the Python LSP server, this shows up as text wherever this property is used: 

<img width="739" height="191" alt="Image" src="https://github.com/user-attachments/assets/af254c4a-bb0d-45cf-9c28-a34cdf544a9a" />

I figured it'd make sense to do this rather than a normal comment because we use this in a way that's not necessarily intuitive from just the property name — this way the additional information shows up wherever it's used.

tw4l replied:
Oh cool! Based on this pydantic issue, it seems maybe it should be wrapped in triple quotes as a docstring?: https://github.com/pydantic/pydantic/issues/5656

But admittedly this is new to me so I may well be wrong!

emma-sg replied:
Oh nice, yeah — just requires `model_config = ConfigDict(use_attribute_docstrings=True)` to be set for that to work.

tw4l commented on backend/btrixcloud/subs.py:
```diff
@@ -182,6 +184,51 @@ async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, boo
+
+        if not org.subscription.futureCancelDate:
+            print(f"Future cancel date not found for subscription ID {reminder.subId}")
+            raise HTTPException(status_code=500, detail="future_cancel_date_not_found")
```
```suggestion
            raise HTTPException(status_code=400, detail="future_cancel_date_not_found")
```


Does cashew expect a 500? Otherwise, we typically use 400 for exceptions like these.

emma-sg replied:
Nope! Can swap to 400.

tw4l commented on backend/btrixcloud/subs.py:
```diff
@@ -182,6 +184,51 @@ async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, boo
+
+        if len(users) == 0:
+            print(f"No admin users found for organization ID {org.id}")
+            raise HTTPException(status_code=500, detail="no_admin_users_found")
```
```suggestion
            raise HTTPException(status_code=400, detail="no_admin_users_found")
```

tw4l commented on backend/btrixcloud/subs.py:
```diff
@@ -182,6 +184,51 @@ async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, boo
+                f"Subscription not found for organization ID {org.id} with sub id {reminder.subId}"
+            )
+            raise HTTPException(status_code=500, detail="subscription_not_found")
+
```
```suggestion
```

Since we just fetched the org by its `subscription.subId` value equalling `reminder.subId` above, probably safe to assume the subscription is set on the org here.

emma-sg replied:
Sure yeah, mostly just did this to tell the type checker that `org.subscription` is not `None`. I can just `assert` it instead though.

tw4l replied:
Ah, fair enough. Feel free to keep and just swap to a `400` if you'd prefer :)

tw4l reviewed (commented):
Tested on dev, working well. Left a few small suggestions.

emma-sg committed:
swap to 400 error codes when sending subscription reminder email

also swaps the more explicit org subscription check, since we just
fetched an org by its subscription id

emma-sg committed:
enable `use_attribute_docstrings` in Subscription

this allows a docstring to show up both in LSP outputs & in OpenAPI
outputs

emma-sg committed:
update trial end text for when user has cancelled trial

emma-sg committed:
add "read-only" state for trial end emails

to be used when a user has a subscription without a payment method

emma-sg committed:
update trial cancellation text when user has cancelled manually

emma-sg committed:
include commit hash with email service logs

emma-sg committed:
simplify & use commit hash from github action env

emma-sg committed:
remove script from package.json

ikreymer commented on emails/emails/trial-ending-soon.tsx:
```diff
@@ -117,8 +117,7 @@ export const TrialEndingSoonEmail = ({
           <Warning>
-            If you opt to end your trial without continuing, all data hosted on
-            Browsertrix under the{" "}
+            Otherwise, all data hosted on Browsertrix under the{" "}
```
```suggestion
            All data hosted on Browsertrix under the{" "}
```

ikreymer committed:
Apply suggestion from @ikreymer

ikreymer commented on backend/btrixcloud/subs.py:
```diff
@@ -182,6 +184,47 @@ async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, boo
+            ]
+        )
+
+        return SubscriptionReminderResponse(sent=True)
```
Small nitpick: I think should just use `SuccessResponse(success=True)` rather than adding a new API response type?

emma-sg replied:
Oh sure, yeah

ikreymer commented on backend/btrixcloud/subs.py:
```diff
@@ -395,6 +438,17 @@ async def cancel_subscription(
         return await ops.cancel_subscription(cancel)
 
+    @app.post(
+        "/subscriptions/trial-end-reminder",
```
```suggestion
        "/subscriptions/send-trial-end-reminder",
```
Maybe should rename this for clarity that all this does is send the reminder?

ikreymer committed:
tweaks:
- resuse SuccessResponse instead of adding new model
- change endpoint to /send-trial-end-reminder for clarity
